### PR TITLE
Use ExceptionDispatchInfo to retain call stack in Session.WaitOnHandle()

### LINKED ISSUE #935: Session.WaitOnHandle loosing call stack
I have seen a number of issues where the call stack provided does not reflect the real location of the exception, making diagnosis nearly impossible.
This is caused by ```Session.WaitOnHandle``` using ```throw _exception;```.

### PR DESCRIPTION
**Pull request by IgorMilavec**
This will fix #935 on newer frameworks

### CODE REVIEW

**Comment by IgorMilavec** [comment]
@drieseng , can you have a look at this? It will help us in getting better information to diagnose problems in Session.

**Review by Rob-Hague — APPROVED** [review]
@WojciechNagorski This change is very useful

Example stack trace before:

```
Test method Renci.SshNet.IntegrationTests.HostKeyAlgorithmTests.SshRsaSha512 threw exception: 
System.NullReferenceException: Object reference not set to an instance of an object.
TestCleanup method Renci.SshNet.IntegrationTests.HostKeyAlgorithmTests.TearDown threw exception. System.NullReferenceException: Object reference not set to an instance of an object..

  Stack Trace: 
Session.WaitOnHandle(WaitHandle waitHandle, TimeSpan timeout) line 995
Session.WaitOnHandle(WaitHandle waitHandle) line 965
Session.Connect() line 641
BaseClient.CreateAndConnectSession() line 533
BaseClient.Connect() line 237
HostKeyAlgorithmTests.DoTest(HostKeyAlgorithm hostKeyAlgorithm, HostKeyFile hostKeyFile, Int32 keyLength) line 73
HostKeyAlgorithmTests.SshRsaSha512() line 50
RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
```

After:

```
  Message: 
Test method Renci.SshNet.IntegrationTests.HostKeyAlgorithmTests.SshRsaSha512 threw exception: 
System.NullReferenceException: Object reference not set to an instance of an object.
TestCleanup method Renci.SshNet.IntegrationTests.HostKeyAlgorithmTests.TearDown threw exception. System.NullReferenceException: Object reference not set to an instance of an object..

  Stack Trace: 
RsaKey.get_D() line 51
RsaCipher.ctor(RsaKey key) line 25
RsaDigitalSignature.ctor(RsaKey rsaKey, HashAlgorithmName hashAlgorithmName) line 29
<.ctor>b__132_25(Byte[] data) line 409
KeyExchangeEC.ValidateExchangeHash() line 81
KeyExchange.Finish() line 161
KeyExchangeECCurve25519.Finish() line 61
KeyExchangeECCurve25519.Session_KeyExchangeEcdhReplyMessageReceived(Object sender, MessageEventArgs`1 e) line 91
Session.OnKeyExchangeEcdhReplyMessageReceived(KeyExchangeEcdhReplyMessage message) line 1410
KeyExchangeEcdhReplyMessage.Process(Session session) line 69
<24 more frames...>
Session.WaitOnHandle(WaitHandle waitHandle, TimeSpan timeout) line 996
Session.WaitOnHandle(WaitHandle waitHandle) line 966
Session.Connect() line 642
BaseClient.CreateAndConnectSession() line 533
BaseClient.Connect() line 237
RemoteSshdConfig.Update() line 228
RemoteSshdConfigExtensions.Reset(RemoteSshdConfig remoteSshdConfig) line 11
HostKeyAlgorithmTests.TearDown() line 23
RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
```

**Review by WojciechNagorski — COMMENTED** [review]
Merge with master

**Review by WojciechNagorski — COMMENTED** [review]
Merge

**Review by WojciechNagorski — APPROVED** [review]

File: src/Renci.SshNet/Session.cs [hunk 1]
```diff
@@ -992,6 +992,7 @@ internal void WaitOnHandle(WaitHandle waitHandle, TimeSpan timeout)
             switch (signaledElement)
             {
                 case 0:
+                    System.Runtime.ExceptionServices.ExceptionDispatchInfo.Capture(_exception).Throw();
                     throw _exception;
```
Comment (Rob-Hague) [inline src/Renci.SshNet/Session.cs]:
```suggestion
                    break;
```

Comment (Rob-Hague) [reply src/Renci.SshNet/Session.cs]: Feel free to ignore that 🙂

Comment (WojciechNagorski) [reply src/Renci.SshNet/Session.cs]: I was thinking the same thing, even though it's unused code anyway.

**Comment by Rob-Hague** [comment]
@WojciechNagorski @IgorMilavec Thanks!

**Pull request merged**